### ISSUE TITLE
Inference time decreases as the input size increases

### ISSUE THREAD

**Issue by PakoAzarashi** [ultralytics/ultralytics#20446]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

During validation, I tested multiple model scales and input sizes. Interestingly, I found that for a given model scale, increasing the input size did not lead to a longer inference time.

Is this performance considered normal?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PakoAzarashi, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will review your question and assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for more details on model performance, as well as [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage guides, where common questions are often addressed.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the behavior you described.

If your question is related to custom training or validation, please share additional context such as dataset examples, validation commands, and relevant logs. Also, review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal setup.

For real-time discussion, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth topics, visit [Discourse](https://community.ultralytics.com/), or connect with the community on our [Subreddit](https://reddit.com/r/Ultralytics)!

## Upgrade

Make sure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your detailed report! Your question will be reviewed by our team soon.

**Comment by Y-T-G** [comment]
What's the code you used to test?

**Comment by PakoAzarashi** [comment]
```
import pandas as pd
import openpyxl
import os
from ultralytics.utils import torch_utils
from ultralytics import YOLO
import itertools
import numpy as np
import time

t = time.time()
date = time.strftime('%Y%m%d')

def check_file(output_file):
    if os.path.isfile(output_file):
        os.remove(output_file)

def get_vals(model_path, imgsz, save_name, yaml_path):
    # Load a model
    model = YOLO(model_path)
    # Customize validation settings
    validation_results = model.val(data=yaml_path,
                                split='val', 
                                imgsz=imgsz, 
                                batch=1, 
                                conf=0.5, 
                                iou=0.5, 
                                device="0",
                                name=save_name,
                                )
    
    mAP50 = validation_results.box.map50
    mAP5095 = round(validation_results.box.map, 5)

    AP50_beewith,  AP50_beewithout = validation_results.box.ap50
    AP5095_beewith,  AP5095_beewithout = np.round(validation_results.box.ap, 5)

    Reacll50_bee = round(validation_results.confusion_matrix.matrix[[0,1], [0,1]].sum()/validation_results.confusion_matrix.matrix.sum(0)[0:2].sum(), 4)
    Precision50_bee = round(validation_results.confusion_matrix.matrix[[0,1], [0,1]].sum()/validation_results.confusion_matrix.matrix.sum(), 4)

    Reacll50_beewith, Reacll50_beewithout = np.round(validation_results.box.r, 4)
    Precision50_beewith, Precision50_beewithout = np.round(validation_results.box.p, 4)

    inference_time = round(validation_results.speed['inference'],2)
    all_time = round(sum(list(validation_results.speed.values())),2)

    Parameters_M = torch_utils.get_num_params(model)/1000**2
    FLOPS_G = torch_utils.get_flops_with_torch_profiler(model=model, imgsz=imgsz)
    FPS = round(1000 / inference_time, 2)
    result_df = pd.DataFrame([{
            'Model': save_name,
            'mAP50': mAP50,
            'mAP5095': mAP5095,
            'AP50_beewith': AP50_beewith,
            'AP5095_beewith': AP5095_beewith,
            'AP50_beewithout': AP50_beewithout,
            'AP5095_beewithout': AP5095_beewithout,
            'Recall50_bee': Reacll50_bee,
            'Precision50_bee': Precision50_bee,
            'Recall50_beewith': Reacll50_beewith,
            'Recall50_beewithout': Reacll50_beewithout,
            'Precision50_beewith': Precision50_beewith,
            'Precision50_beewithout': Precision50_beewithout,
            'Inference Time (ms)': inference_time,
            'All Time (ms)': all_time,
            "Parameters (M) ": Parameters_M,
            "FLOPS (G)":FLOPS_G,
            "FPS": FPS
        }])
    return result_df
```


this is the main code, and I tried n, s, m, l model scales and  512, 576, 640, 704, 768 input sizes

**Comment by Y-T-G** [comment]
The decrease you see is probably due to caching since you're running validation on the same dataset. For better estimates, you should use a large dataset (1000+ images).

**Comment by Swaghaug** [comment]
You should verify that Ultralytics doesn't scale your images before running inference on them.

If you are only timing your inference time and not the preprocessing step of resizing to the default (640) you are essentially timing identical operations.

Try timing a predict function on a single image and very the image sizes to verify that your results still occur.

**Comment by glenn-jocher** [comment]
You're observing an interesting behavior that can be explained by several factors. When using the validation method, GPU caching effects and batch processing optimizations can mask the expected relationship between input size and inference time. 

For more accurate benchmarking, I recommend:

1. Using the `benchmark()` function which is specifically designed for consistent performance measurement: `from ultralytics.utils.benchmarks import benchmark`

2. Timing direct prediction on single images of different sizes using:
```python
model = YOLO("yolo11n.pt")
for size in [512, 640, 768]:
    img = np.zeros((size, size, 3), dtype=np.uint8)  # Create blank image
    start = time.time()
    model.predict(img, verbose=False)
    print(f"Size {size}x{size}: {(time.time() - start) * 1000:.2f}ms")
```

This will eliminate potential caching effects and provide a clearer picture of how input size affects inference speed.

**Comment by PakoAzarashi** [comment]
Thank you guys so much for responses. And I am grateful for your support and consideration.

I tried the solutions, got the increase inference time while increase the input size.
Thank you guys so much

**Comment by glenn-jocher** [comment]
You're welcome, @PakoAzarashi! I'm glad to hear you were able to observe the expected relationship between input size and inference time using the suggested approaches. This is indeed the normal behavior - larger input sizes require more processing time. The benchmark function is always helpful for these kinds of performance evaluations when you need consistent results. Feel free to reach out if you have any other questions!
